### PR TITLE
Add Differentiable CEM solver

### PR DESCRIPTION
**Pull request by dishank-b**
## Motivation and Context

This adds a differential CEM solver to the optimizers class of Theseus. Can be used to solve control problems. 
Based on https://github.com/facebookresearch/dcem

## How Has This Been Tested

Passed the unit tests given. 

## Types of changes

- [x] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## ToDos

- [x] Check with `theseus_layer_test` provided
- [ ] Run with different backwards modes
- [x] Validate the gradients through the optimizer

### CODE REVIEW

File: theseus/optimizer/nonlinear/dcem.py [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+        info.status[
+            info.status == NonlinearOptimizerStatus.START
+        ] = NonlinearOptimizerStatus.MAX_ITERATIONS
+
+        self.objective.update(self._mu_vec_to_dict(mu))
+
+        # return self._mu_vec_to_dict(mu)
+        return itr
+
+    def _solve(self):
```
Comment (bamos) [inline theseus/optimizer/nonlinear/dcem.py:309]: Can maybe rename to `_CEM_step` and switch to passing the previous sigma as an input?

File: theseus/optimizer/nonlinear/dcem.py [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+
+    def _reinit_sigma(self, init_sigma=1.0, **kwargs):
+        self.sigma = (
+            torch.ones(
+                (self.ordering[0].shape[0], self.tot_dof), device=self.objective.device
+            )
+            * init_sigma
+        )
+
+    def _all_solve(
```
Comment (bamos) [inline theseus/optimizer/nonlinear/dcem.py:192]: Can this be deleted?

**Review by bamos — CHANGES_REQUESTED** [review]
It looks like a nice initial version! I left a few minor comments inline to the core parts of the code and will review the notebooks in some more detail soon. And how are the tests looking?

File: theseus/__init__.py [hunk 3]
```diff
@@ -84,6 +84,7 @@
 from .optimizer.nonlinear import (  # usort: skip
     BackwardMode,
     GaussNewton,
+    DCem,
```
[the remark this answers is not in the document]
Comment (dishank-b) [reply theseus/__init__.py:87]: yeah, will do

File: theseus/optimizer/nonlinear/dcem.py (hunk 2 above)
Comment (dishank-b) [reply theseus/optimizer/nonlinear/dcem.py:192]: yeah

File: theseus/optimizer/nonlinear/dcem.py (hunk 1 above)
Comment (dishank-b) [reply theseus/optimizer/nonlinear/dcem.py:309]: yeah

**Comment by dishank-b** [comment]
@luisenp @mhmukadam . Can you also have a look at this. I think it's ready to be merged from my side, a few other things can be done but that can be done in later PRs.

File: tutorials/06_DCEM.ipynb
Comment (luisenp) [inline tutorials/06_DCEM.ipynb:1]: What are the specific differences between this and tutorial 2? Is it just the optimizer class that changes, or does some of the other code needed to change as well?

File: tests/optimizer/nonlinear/test_dcem.py
Comment (luisenp) [inline tests/optimizer/nonlinear/test_dcem.py:1]: Do we remove this file?

**Review by luisenp — CHANGES_REQUESTED** [review]
This looks great! Left a bunch of minor comments that should be easy to address. Thanks!

File: tests/test_theseus_layer.py [hunk 4]
```diff
@@ -85,7 +85,7 @@ def create_qf_theseus_layer(
     cost_weight=th.ScaleCostWeight(1.0),
     nonlinear_optimizer_cls=th.GaussNewton,
     linear_solver_cls=th.CholeskyDenseSolver,
-    max_iterations=10,
+    max_iterations=50,
```
Comment (luisenp) [inline tests/test_theseus_layer.py:88]: Why change this default?

File: theseus/optimizer/nonlinear/__init__.py [hunk 5]
```diff
@@ -2,7 +2,9 @@
 #
 # This source code is licensed under the MIT license found in the
 # LICENSE file in the root directory of this source tree.
+
 from .dogleg import Dogleg
+from .dcem import DCEM
```
Comment (luisenp) [inline theseus/optimizer/nonlinear/__init__.py:7]: nit: Alphabetical order.

File: theseus/optimizer/nonlinear/dcem.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+
+    def _mu_vec_to_dict(self, mu):
+        idx = 0
+        mu_dic = {}
+        for var in self.ordering:
+            mu_dic[var.name] = mu[:, slice(idx, idx + var.dof())]
+            idx += var.dof()
+        return mu_dic
+
+    def _reinit_sigma(self, init_sigma):
```
Comment (luisenp) [inline theseus/optimizer/nonlinear/dcem.py:84]: Maybe better called `reset_sigma`.

File: setup.py [hunk 7]
```diff
@@ -46,7 +53,10 @@ def parse_requirements_file(path):
                 # installed
                 continue
             line = line.strip()
-            reqs.append(line)
+            if line[0] == "-":
```
[the remark this answers is not in the document]
Comment (luisenp) [reply setup.py:56]: Let's remove it, since it's not used anymore. We can add later if necessary, but I also try to avoid installing directly from github unless there is no other choice.

File: tests/test_theseus_layer.py (hunk 4 above)
Comment (luisenp) [reply tests/test_theseus_layer.py:88]: Yeah, better to set the value in the function call rather than change the default (which might silently affect a lot of other code).

File: theseus/optimizer/nonlinear/dcem.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+from theseus.optimizer.variable_ordering import VariableOrdering
+
+from .nonlinear_optimizer import (
+    NonlinearOptimizer,
+    BackwardMode,
+    NonlinearOptimizerInfo,
+    NonlinearOptimizerStatus,
+)
+
+EndIterCallbackType = Callable[
```
[the remark this answers is not in the document]
Comment (luisenp) [reply theseus/optimizer/nonlinear/dcem.py:19]: I think it's fine to do it in this PR, it's not a big change.

File: tests/optimizer/nonlinear/test_dcem.py
Comment (luisenp) [reply tests/optimizer/nonlinear/test_dcem.py:1]: Should we set up a meeting for this? Are a lot of changes necessary?

File: theseus/optimizer/nonlinear/dcem.py [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+                indexes = torch.softmax(-_fX * self.temp, dim=1)
+            else:
+                indexes = LML(
+                    N=self.n_elite, verbose=self.lml_verbose, eps=self.lml_eps
+                )(-_fX * self.temp)
+            indexes = indexes.unsqueeze(2)
+
+        else:
+            indexes_vals = fX.argsort(dim=1)[:, : self.n_elite]
+            # TODO: A scatter would be more efficient here.
```
[the remark this answers is not in the document]
Comment (luisenp) [reply theseus/optimizer/nonlinear/dcem.py:134]: Ah ok, didn't really pay attention to the condition, but this is for  temp = $\infty$, which I guess is an uncommon case? Maybe give it a quick shot, but if it gets messy feel free to ignore.

**Review by luisenp — COMMENTED** [review]
Thanks for addressing these. Left a couple of follow up questions.

File: tests/test_theseus_layer.py (hunk 4 above)
Comment (dishank-b) [reply tests/test_theseus_layer.py:88]: make sense, will revert

File: theseus/optimizer/nonlinear/dcem.py [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+        self.n_elite = n_elite
+        self.lb = lb
+        self.ub = ub
+        self.temp = temp
+        self.normalize = normalize
+        self.tot_dof = sum([x.dof() for x in self.ordering])
+        self.lml_eps = lml_eps
+        self.lml_verbose = lml_verbose
+        self.init_sigma = init_sigma
+
```
[the remark this answers is not in the document]
Comment (dishank-b) [reply theseus/optimizer/nonlinear/dcem.py:75]: Yeah, I guess. But info seems to have specific attributes about convergence, errors, history, etc. And since other optimizer also use info class and they don't anything returning like sigma, I am not sure if info is the right place? On the other hand, info makes sense, but maybe adding new attribute such as `auxiliary_info` which is empty by default, but optimizer can add any info they want to return through this? 
May be can keep this for different PR.

File: theseus/optimizer/nonlinear/dcem.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dishank-b) [reply theseus/optimizer/nonlinear/dcem.py:19]: okay

File: theseus/optimizer/nonlinear/dcem.py [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+                    print(
+                        f"Nonlinear optimizer. Iteration: {it_+1}. "
+                        f"Error: {err.mean().item()} "
+                    )
+                converged_indices = self._check_convergence(err, info.last_err)
+                info.status[
+                    np.array(converged_indices.cpu().numpy())
+                ] = NonlinearOptimizerStatus.CONVERGED
+
+                # TODO
```
[the remark this answers is not in the document]
Comment (dishank-b) [reply theseus/optimizer/nonlinear/dcem.py:190]: Actually never mind this, it's working fine. I removed the comment itself

File: tests/optimizer/nonlinear/test_dcem.py
Comment (dishank-b) [reply tests/optimizer/nonlinear/test_dcem.py:1]: It's mainly related to assumption of linearization in the optimizers. Can setup a meeting.

**Review by dishank-b — COMMENTED** [review]
Made the changes, have a look.

File: tests/optimizer/nonlinear/test_dcem.py
Comment (luisenp) [reply tests/optimizer/nonlinear/test_dcem.py:1]: Let's discuss in a meeting. I'm OK with leaving this test for a separate PR (but let's open up an issue for it).

File: theseus/optimizer/nonlinear/dcem.py [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+            indexes = indexes.unsqueeze(2)
+            eps = 0
+
+        else:
+            indexes_vals = fX.argsort(dim=1)[:, : self.n_elite]
+            # Scatter 1.0 to the indexes using indexes_vals
+            indexes = torch.zeros(n_batch, self.n_samples, device=device).scatter_(
+                1, indexes_vals, 1.0
+            )
+            # indexes = torch.zeros(n_batch, self.n_samples, device=device)
```
Comment (luisenp) [inline theseus/optimizer/nonlinear/dcem.py:136]: Can remove the code that is commented out.

**Review by luisenp — APPROVED** [review]
LGTM! Left a couple of comments of things we can move to a separate PR. I think we could also rename some of the args/kwargs, but I'm OK with merging as is. I can take a look at this after we have a unit test in place. Thanks for your hard work on this!

File: tutorials/06_DCEM.ipynb
Comment (dishank-b) [reply tutorials/06_DCEM.ipynb:1]: yeah, sounds good.

File: tests/optimizer/nonlinear/test_dcem.py
Comment (dishank-b) [reply tests/optimizer/nonlinear/test_dcem.py:1]: Yeah, having separate PR makes sense. I will open an issue once merged?

File: theseus/optimizer/nonlinear/dcem.py (hunk 12 above)
Comment (dishank-b) [reply theseus/optimizer/nonlinear/dcem.py]: done
